pkg/agent/route/route_windows.go

```diff
@@ -100,21 +113,42 @@ func (c *Client) Reconcile(podCIDRs []string) error {
 // It overrides the routes if they already exist, without error.
 func (c *Client) AddRoutes(podCIDR *net.IPNet, nodeName string, peerNodeIP, peerGwIP net.IP) error {
 	obj, found := c.hostRoutes.Load(podCIDR.String())
+	route := &netroute.Route{
+		DestinationSubnet: podCIDR,
+	}
+	if c.networkConfig.TrafficEncapMode.NeedsEncapToPeer(peerNodeIP, c.nodeConfig.NodeIPAddr) {
+		route.LinkIndex = c.nodeConfig.GatewayConfig.LinkIndex
+		route.GatewayAddress = peerGwIP
+	} else if !c.networkConfig.TrafficEncapMode.NeedsRoutingToPeer(peerNodeIP, c.nodeConfig.NodeIPAddr) {
+		// NoEncap traffic to Node on the same subnet.
+		// Set the peerNodeIP as next hop.
+		route.LinkIndex = c.bridgeInfIndex
+		route.GatewayAddress = peerNodeIP
+	}
+	// NoEncap traffic to Node on the different subnet needs underlying routing support.
+	// Use host default route inside the Node.
+
 	if found {
-		rt := obj.(*netroute.Route)
-		if rt.GatewayAddress.Equal(peerGwIP) {
+		existingRoute := obj.(*netroute.Route)
+		if existingRoute.GatewayAddress.Equal(route.GatewayAddress) {
 			klog.V(4).Infof("Route with destination %s already exists on %s (%s)", podCIDR.String(), nodeName, peerNodeIP)
 			return nil
 		}
 		// Remove the existing route entry if the gateway address is not as expected.
-		if err := c.nr.RemoveNetRoute(rt.LinkIndex, rt.DestinationSubnet, rt.GatewayAddress); err != nil {
+		if err := c.nr.RemoveNetRoute(existingRoute.LinkIndex, existingRoute.DestinationSubnet, existingRoute.GatewayAddress); err != nil {
 			klog.Errorf("Failed to delete existing route entry with destination %s gateway %s on %s (%s)", podCIDR.String(), peerGwIP.String(), nodeName, peerNodeIP)
 			return err
 		}
 	}
-	if err := c.nr.NewNetRoute(c.nodeConfig.GatewayConfig.LinkIndex, podCIDR, peerGwIP); err != nil {
+
+	if route.GatewayAddress == nil {
+		return nil
+	}
+
+	if err := c.nr.NewNetRoute(route.LinkIndex, route.DestinationSubnet, route.GatewayAddress); err != nil {
 		return err
 	}
+
 	c.hostRoutes.Store(podCIDR.String(), &netroute.Route{
 		LinkIndex:         c.nodeConfig.GatewayConfig.LinkIndex,
```

Comment:
Shouldn't the stored route be updated too? otherwise the method will keep removing the route and add new one? could you check logs?

Reply:
Will check logs.

Reply:
I think so. `NewNetRoute` takes values in `route` and `route` should be stored. Updated.
However, I checked logs and there's no extra logs like `Added route with destination...` in antrea windows log.. It seems removing and adding repeatly doesn't happen. It is unexpected.

Reply:
Was it because no Node event triggers it? Maybe you can try annotating the Node

Reply:
I added some logs and found that actually the method returns here: https://github.com/antrea-io/antrea/blob/fa8a1974db49cf93be11ad454f49842f5bde6108/pkg/agent/route/route_windows.go#L135

Regardless of the change, the gateway IP is always stored so new code doesn't introduce repeated route addition.

Now, I have updated the code to store `route`.

Reply:
It doesn't make sense then. The GatewayAddress in the stored route is always `peerGwIP`, but the desired GatewayAddress is `peerNodeIP`, they shouldn't equal. Did you test encap mode or noEncap mode?

Reply:
I think I was talking about encap mode for the comments above.

Reply:
That should be result of encap mode with old code. Maybe I didn't clean up routes before testing but I don't remember clearly.
I just tested with latest code, encap and noecap. Then I checked logs & routes and they are expected.